##1. Bibliotecas necessárias

In [ ]:
import pip
import pandas as pd

In [ ]:
!pip install dask[complete]

     |████████████████████████████████| 830 kB 2.4 MB/s 
     |████████████████████████████████| 133 kB 51.2 MB/s 
     |████████████████████████████████| 822 kB 49.8 MB/s 
     |████████████████████████████████| 802 kB 44.7 MB/s 
     |████████████████████████████████| 802 kB 36.2 MB/s 
     |████████████████████████████████| 793 kB 51.7 MB/s 
     |████████████████████████████████| 793 kB 39.5 MB/s 
     |████████████████████████████████| 791 kB 37.8 MB/s 
     |████████████████████████████████| 786 kB 37.9 MB/s 
     |████████████████████████████████| 779 kB 40.0 MB/s 
     |████████████████████████████████| 778 kB 42.7 MB/s 
     |████████████████████████████████| 776 kB 36.3 MB/s 
     |████████████████████████████████| 769 kB 40.1 MB/s 
     |████████████████████████████████| 766 kB 44.8 MB/s 
     |████████████████████████████████| 1.0 MB 39.7 MB/s 
     |████████████████████████████████| 722 kB 43.1 MB/s 
     |████████████████████████████████| 722 kB 34.9 MB/s 
     |█████████

In [ ]:
#importando o dask 
import dask.array as da
import dask.dataframe as dd

Cria dask array de dimensoes de 10000 por 10000

##2. Importando base de dados para analise


In [ ]:
#acessando drive
from google.colab import drive

drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
##lendo a base de dados que esta quase completa
df = dd.read_csv('/content/drive/MyDrive/Colab Notebooks/IC/eccomerce/database/dataframes/df2_completo.csv')

In [ ]:
#verificando as infos
df = df.compute()
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 119143 entries, 0 to 674
Data columns (total 39 columns):
 #   Column                         Non-Null Count   Dtype  
---  ------                         --------------   -----  
 0   order_id                       119143 non-null  object 
 1   order_item_id                  118310 non-null  float64
 2   product_id                     118310 non-null  object 
 3   seller_id                      118310 non-null  object 
 4   shipping_limit_date            118310 non-null  object 
 5   price                          118310 non-null  float64
 6   freight_value                  118310 non-null  float64
 7   payment_sequential             119140 non-null  float64
 8   payment_type                   119140 non-null  object 
 9   payment_installments           119140 non-null  float64
 10  payment_value                  119140 non-null  float64
 11  review_id                      118146 non-null  object 
 12  review_score                   11

##3. Trabalhando a base de dados geolocation

In [ ]:
#importando a base com geolocalizações
df_geolocation = pd.read_csv("/content/drive/MyDrive/Colab Notebooks/IC/eccomerce/database/dataframes/olist_geolocation_dataset.csv")

In [ ]:
df_geolocation.head(10)

geolocation_zip_code_prefix  ...  geolocation_state
0                         1037  ...                 SP
1                         1046  ...                 SP
2                         1046  ...                 SP
3                         1041  ...                 SP
4                         1035  ...                 SP
5                         1012  ...                 SP
6                         1047  ...                 SP
7                         1013  ...                 SP
8                         1029  ...                 SP
9                         1011  ...                 SP

[10 rows x 5 columns]

para um mesmo codigo postal tem-se lat e long diferentes, como por exemplo ali no 1046
vamos fazer entao a media para um mesmo codigo postal

###3.1 Verificando dados faltantes

####3.1.1 Teste com pandas 

In [ ]:
#teste
import pandas as pd
teste_geo = pd.read_csv("/content/drive/MyDrive/Colab Notebooks/IC/eccomerce/database/dataframes/olist_geolocation_dataset.csv")

In [ ]:
teste_geo.value_counts()

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  geolocation_city  geolocation_state
88220                        -27.102099       -48.629613       itapema           SC                   314
6414                         -23.495901       -46.874687       barueri           SP                   189
                             -23.490618       -46.869004       barueri           SP                   127
5145                         -23.506049       -46.717377       sao paulo         SP                   126
22620                        -23.005514       -43.375964       rio de janeiro    RJ                   102
                                                                                                     ... 
39400                        -16.729617       -43.869854       montes claros     MG                     1
                             -16.729663       -43.865562       montes claros     MG                     1
                             -16.729720       -43.864

####3.1.2 Dados nulos


In [ ]:
#utilizando o drop_duplicates, retorna um novo dataframe sem as duplicações
new_df_geo = df_geolocation.drop_duplicates()

In [ ]:
new_df_geo.head(15)

geolocation_zip_code_prefix  ...  geolocation_state
0                          1037  ...                 SP
1                          1046  ...                 SP
2                          1046  ...                 SP
3                          1041  ...                 SP
4                          1035  ...                 SP
5                          1012  ...                 SP
6                          1047  ...                 SP
7                          1013  ...                 SP
8                          1029  ...                 SP
9                          1011  ...                 SP
10                         1013  ...                 SP
11                         1032  ...                 SP
12                         1014  ...                 SP
13                         1012  ...                 SP
14                         1037  ...                 SP

[15 rows x 5 columns]

Agora, faz-se a media da latitude e longitude para o mesmo cep

In [ ]:
#utilizando groupby
new_df_geo = new_df_geo.groupby(by="geolocation_zip_code_prefix").mean()


In [ ]:
new_df_geo.head(15)

geolocation_lat  geolocation_lng
geolocation_zip_code_prefix                                  
1001                              -23.550227       -46.634039
1002                              -23.547657       -46.634991
1003                              -23.549000       -46.635582
1004                              -23.549829       -46.634792
1005                              -23.549547       -46.636406
1006                              -23.550127       -46.636044
1007                              -23.549962       -46.637204
1008                              -23.546000       -46.635877
1009                              -23.546891       -46.636498
1010                              -23.546667       -46.635404
1011                              -23.546649       -46.635425
1012                              -23.547808       -46.634844
1013                              -23.547240       -46.634221
1014                              -23.546040       -46.633600
1015                              -23.547756       -46.631738

Para cada vendedor e para cada cliente ter um cep associado

In [ ]:
#Identificando o cliente
#left para o df completo e rigth para o com as localizações
costumer = df.merge(new_df_geo, left_on='customer_zip_code_prefix', right_on='geolocation_zip_code_prefix')

In [ ]:
#certo, cada cliente tem um lat e um long
costumer.head(5)

order_id  ...  geolocation_lng
0  00010242fe8c5a6d1ba2dd792cb16214  ...       -41.310265
1  bd31b009e1dbc47fc7c250b1e2cf5440  ...       -41.310265
2  bb4eb0196897c20281a61f75ce23211c  ...       -41.310265
3  f01cc1e4fe1833e42f696b0dcb0d77ef  ...       -41.310265
4  f4776a61a23801c349ab69bd0fd774e2  ...       -41.310265

[5 rows x 41 columns]

Problema: agora tem-se duas colunas com as informações de geolocalizacao
retira-se as erradas: costumer_lat e costumer_long

In [ ]:
#a coluna para latitude do cliente recebe o valor correto
costumer['costumer_lat'] = costumer['geolocation_lat']

In [ ]:
#apaga a variavel geolocation que nao precisa mais
costumer = costumer.drop(['geolocation_lat'],axis=1)

In [ ]:
#a coluna para longitude do cliente recebe o valor correto
costumer['costumer_lng'] = costumer['geolocation_lng']

In [ ]:
#apaga a variavel geolocation que nao precisa mais
costumer = costumer.drop(['geolocation_lng'],axis=1)

In [ ]:
costumer.head(5)

order_id  order_item_id  ... costumer_lat costumer_lng
0  00010242fe8c5a6d1ba2dd792cb16214            1.0  ...   -21.763186   -41.310265
1  bd31b009e1dbc47fc7c250b1e2cf5440            1.0  ...   -21.763186   -41.310265
2  bb4eb0196897c20281a61f75ce23211c            1.0  ...   -21.763186   -41.310265
3  f01cc1e4fe1833e42f696b0dcb0d77ef            1.0  ...   -21.763186   -41.310265
4  f4776a61a23801c349ab69bd0fd774e2            1.0  ...   -21.763186   -41.310265

[5 rows x 41 columns]

Para o vendedor

In [ ]:
seller = costumer.merge(new_df_geo, left_on='seller_zip_code_prefix', right_on='geolocation_zip_code_prefix')

In [ ]:
#apaga a variavel geolocation que nao precisa mais
#a coluna para latitude do vendedor recebe o valor correto
seller['seller_lat'] = seller['geolocation_lat']
seller = seller.drop(['geolocation_lat'],axis=1)

In [ ]:
#apaga a variavel geolocation que nao precisa mais
#a coluna para long do vendedor recebe o valor correto
seller['seller_lng'] = seller['geolocation_lng']
seller = seller.drop(['geolocation_lng'],axis=1)

In [ ]:
seller.columns

Index(['order_id', 'order_item_id', 'product_id', 'seller_id',
       'shipping_limit_date', 'price', 'freight_value', 'payment_sequential',
       'payment_type', 'payment_installments', 'payment_value', 'review_id',
       'review_score', 'review_comment_title', 'review_comment_message',
       'review_creation_date', 'review_answer_timestamp', 'customer_id',
       'order_status', 'order_purchase_timestamp', 'order_approved_at',
       'order_delivered_carrier_date', 'order_delivered_customer_date',
       'order_estimated_delivery_date', 'customer_unique_id',
       'customer_zip_code_prefix', 'customer_city', 'customer_state',
       'seller_zip_code_prefix', 'seller_city', 'seller_state',
       'product_category_name', 'product_name_lenght',
       'product_description_lenght', 'product_photos_qty', 'product_weight_g',
       'product_length_cm', 'product_height_cm', 'product_width_cm',
       'costumer_lat', 'costumer_lng', 'seller_lat', 'seller_lng'],
      dtype='object')

In [ ]:
costumer.columns

Index(['order_id', 'order_item_id', 'product_id', 'seller_id',
       'shipping_limit_date', 'price', 'freight_value', 'payment_sequential',
       'payment_type', 'payment_installments', 'payment_value', 'review_id',
       'review_score', 'review_comment_title', 'review_comment_message',
       'review_creation_date', 'review_answer_timestamp', 'customer_id',
       'order_status', 'order_purchase_timestamp', 'order_approved_at',
       'order_delivered_carrier_date', 'order_delivered_customer_date',
       'order_estimated_delivery_date', 'customer_unique_id',
       'customer_zip_code_prefix', 'customer_city', 'customer_state',
       'seller_zip_code_prefix', 'seller_city', 'seller_state',
       'product_category_name', 'product_name_lenght',
       'product_description_lenght', 'product_photos_qty', 'product_weight_g',
       'product_length_cm', 'product_height_cm', 'product_width_cm',
       'costumer_lat', 'costumer_lng'],
      dtype='object')

In [ ]:
#seller = seller.compute()

In [ ]:
#salvando df geolocations prontinho
from google.colab import files
seller.to_csv('olist_geolocated_transactions.csv', index = False)
files.download('olist_geolocated_transactions.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

##4. Verificando se funcionou plotando graficos 

In [ ]:
import plotly.express as px

fig = px.scatter_mapbox(seller, lat = 'seller_lat', lon = 'seller_lng',
                        color_discrete_sequence=["fuchsia"], zoom=3, height=500)
fig.update_layout(mapbox_style="open-street-map")
fig.update_layout(margin={"r":0,"t":0,"l":0,"b":0
fig.show()

In [ ]:
import plotly.express as px
fig = px.density_mapbox(seller, lat='seller_lat', lon='seller_lng', z='price', radius=10,
                        center=dict(lat=0, lon=180), zoom=500,
                        mapbox_style="stamen-terrain")
fig.show()